keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2346 @@
+       tell resource driver to deallocate it by setting `Deallocate` and
+       removing the pod from `ReservedFor` (if present there)>
+    }
+  } else if <all resources allocated> {
+    <schedule pod onto node>
+  } else if <some unallocated resource uses delayed allocation> {
+    <tell resource driver to allocate for the chosen node>
+  }
+}
+```
+
+Randomly picking a node without knowing anything about the resource driver may
+or may not succeed. To narrow the choice of suitable nodes for all claims using
+a certain resource class, a node selector can be specified in that class. That
+selector is static and typically will use labels that determine which nodes may
+have resources available.
+
+To gather information about the current state of resource availability, the
+scheduler creates a PodScheduling object. That object is owned by the pod and
+will either get deleted by the scheduler when it is done with pod scheduling or
+through the garbage collector. In the PodScheduling object, the scheduler posts
+the list of all potential nodes that it was left with after considering all
+other pod constraints and requirements. Resource drivers involved in the
+scheduling of the pod respond by adding which of these nodes currently don't
+have sufficient resources available. The next scheduling attempt is then more
+likely to pick a node for which allocation succeeds.
+
+This scheduling information is optional and does not have to be in sync with
+the current ResourceClaim state, therefore it is okay to store it
+separately. Alternatively, the scheduler may be configured to do node filtering
+through scheduler extenders, in which case the PodScheduling object will not be
+needed.
+
+Allowing the scheduler to trigger allocation in parallel to asking for more
+information was chosen because for pods with a single resource claim, the cost
+of guessing wrong is low: the driver just needs to inform the scheduler to try
+again and provide the additional information.
+
+Additional heuristics are possible without changing the proposed API. For
+example, the scheduler might ask for information and wait a while before
+making a choice. This may be more suitable for pods using many different
+resource claims because for those, allocation may succeed for some claims and
+fail for others, which then may need to go through the recovery flow with
+deallocating one or more claims.
+
+### Resource allocation and usage flow
+
+The following steps shows how resource allocation works for a resource that
+gets defined inline in a Pod. Several of these steps may fail without changing
+the system state. They then must be retried until they succeed or something
+else changes in the system, like for example deleting objects.
+
+* **user** creates Pod with inline ResourceClaimTemplate
+* **resource claim controller** checks ResourceClaimTemplate and ResourceClass,
+  then creates ResourceClaim with Pod as owner
+* if *immediate allocation*:
```

Comment:
It's not exactly a 1:1 match. It might make sense to set up a ResourceClaim independently from a pod, but then still delay allocation until the first pod actually uses it. Other combinations, like "inline" and "immediate" make less sense.

The reason why I am showing "immediate" here is that I wanted to show the complete flow in the middle, without assuming that this is only for inline resources. I could make this more obvious by showing the non-inline case first and then extending it by wrapping the already introduced flow with "create claim" + garbage collection.

Reply:
Just brainstorming: if we only allowed two cases (user-managed ResourceClaim with immediate allocation and inline ResourceClaimTemplate with delayed allocation), how would we distinguish between these two cases? And would it still make sense to create separate ResourceClaim objects for inline resources?

I think there are good arguments in favor of the current approach (primarily the unified allocation tracking and less permissions needed for a resource driver), but it also has downsides (scheduler needs to handle certain additional scenarios).

Reply:
> Other combinations, like "inline" and "immediate" make less sense.

This was my only point here :)  As I looked at the logic I couldn't tell if that was a case that mattered or not.  To put it another way:

```
          |       allocation mode     |
lifetime  |  immediate  |   delayed   |
----------+-------------+-------------+
regular   |   allocate  |             |
claim     |     now     |             |
----------+-------------+             |
inline    | allocate when pod starts  |
claim     |                           |
----------+---------------------------+
```

Reply:
We would still need this `if *immediate allocation*` check here because the ResourceClaim that the pod references might have been created by the user.

This will become clearer when I split this flow:
- first I need to show scheduling and usage of a user-created ResourceClaim + Pod
- then I show how the inline case just adds some steps at the start and end